### PR TITLE
Removing dependencies 3/n

### PR DESCRIPTION
**Pull request by melcz**
Removing guava usage from dexlib2/dexbacked

### CODE REVIEW

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/util/VariableSizeLookaheadIterator.java [hunk 1]
```diff
@@ -30,21 +30,73 @@
...
+        }
+        return tryToComputeNext();
+    }
+
+    private boolean tryToComputeNext() {
+        state = STATE_FAILED; // temporary pessimism
+        next = computeNext();
+        if (state != STATE_DONE) {
+          state = STATE_READY;
+          return true;
```
Comment (sgjesse) [inline dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/util/VariableSizeLookaheadIterator.java:84]: nit: 4-space intentation

File: dexlib2/src/main/java/com/android/tools/smali/util/ChainedIterator.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Combines two iterators into a single iterator. The returned iterator iterates across the
+ * elements in {@code a}, followed by the elements in {@code b}. The source iterators are
+ * not polled until necessary.
+ *
+ * <p>The returned iterator does not support {@code remove()}.
+ */
+public class ChainedIterator<T extends @Nullable Object> implements Iterator<T>, Iterable<T> {
```
Comment (sgjesse) [inline dexlib2/src/main/java/com/android/tools/smali/util/ChainedIterator.java:45]: nit: 4-space intentation for this class.

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java [hunk 3]
```diff
@@ -204,7 +204,9 @@ public static DexBackedDexFile fromInputStream(@Nullable Opcodes opcodes, @Nonnu
             throws IOException {
         DexUtil.verifyDexHeader(is);
 
-        byte[] buf = ByteStreams.toByteArray(is);
+        int fileSize = is.available();
+        byte[] buf = new byte[fileSize];
+        is.read(buf);
```
Comment (sgjesse) [inline dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java:209]: This code pattern is here twice. Maybe add a FileUtils/ByteStreamUtils to util.

**Review by sgjesse — CHANGES_REQUESTED** [review]
Please consider these comments.

Regarding formatting, we could consider running the code through a Java formatter as some point.

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java (hunk 3 above)
Comment (zerny) [reply dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java:209]: InputStream.available does not read the file in full, so this is not equivalent to the full read done by ByteStreams.toByteArray.

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedOdexFile.java [hunk 4]
```diff
@@ -85,14 +84,16 @@ public DexBackedOdexFile(@Nonnull Opcodes opcodes, @Nonnull byte[] odexBuf, byte
         DexUtil.verifyOdexHeader(is);
 
         is.reset();
-        byte[] odexBuf = new byte[OdexHeaderItem.ITEM_SIZE];
-        ByteStreams.readFully(is, odexBuf);
+        byte[] odexBuf = is.readNBytes(OdexHeaderItem.ITEM_SIZE);
         int dexOffset = OdexHeaderItem.getDexOffset(odexBuf);
         if (dexOffset > OdexHeaderItem.ITEM_SIZE) {
-            ByteStreams.skipFully(is, dexOffset - OdexHeaderItem.ITEM_SIZE);
+            is.skip(dexOffset - OdexHeaderItem.ITEM_SIZE);
```
Comment (zerny) [inline dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedOdexFile.java:90]: Same issue for skip here. The java impl may only read a smaller number for bytes. Using readNBytes might be the simplest replacement.

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedOdexFile.java [hunk 5]
```diff
@@ -85,14 +84,16 @@ public DexBackedOdexFile(@Nonnull Opcodes opcodes, @Nonnull byte[] odexBuf, byte
...
         }
 
-        byte[] dexBuf = ByteStreams.toByteArray(is);
+        int remainingBytes = is.available();
+        byte[] dexBuf = new byte[remainingBytes];
+        is.read(dexBuf);
+        is.close();
```
Comment (zerny) [inline dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedOdexFile.java:96]: ditto

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java (hunk 3 above)
Comment (zerny) [reply dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java:209]: I think InputStream.readAllBytes is a valid replacement.

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/ZipDexContainer.java [hunk 6]
```diff
@@ -146,7 +144,9 @@ protected ZipFile getZipFile() throws IOException {
     @Nonnull
     protected DexEntry<DexBackedDexFile> loadEntry(@Nonnull ZipFile zipFile, @Nonnull ZipEntry zipEntry) throws IOException {
         try (InputStream inputStream = zipFile.getInputStream(zipEntry)) {
-            byte[] buf = ByteStreams.toByteArray(inputStream);
+            int fileSize = inputStream.available();
+            byte[] buf = new byte[fileSize];
+            inputStream.read(buf);
```
Comment (zerny) [inline dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/ZipDexContainer.java:149]: ditto

**Review by sgjesse — COMMENTED** [review]
Melisa,

Ian and I just discussed yesterday, the for the inputstream operations from Guava, we should add a library local InputStreamUtil with code compatible with the Guava implementation.

Thinking about this we should also make these implementations without using methods like [readNBytes](https://developer.android.com/reference/java/io/InputStream#readNBytes(int)), as on Android they where not added until API level 33. Maybe smali is not the most obvious library to use on pre API level 33 devices, but there should be no need to break that compatibility right now.

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java (hunk 3 above)
Comment (melcz) [reply dexlib2/src/main/java/com/android/tools/smali/dexlib2/dexbacked/DexBackedDexFile.java]: Included toByteArray, readFully and skipFully to InputStreamUtil

**Comment by melcz** [comment]
Hey, sorry for the commit noise. I included toByteArray, readFully and skipFully in InputStreamUtil and fixed the indentation for newly added files. My bad, the IDE picked it up from importing from the pasted code.

File: dexlib2/src/main/java/com/android/tools/smali/util/InputStreamUtil.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
+
```
Comment (zerny) [inline dexlib2/src/main/java/com/android/tools/smali/util/InputStreamUtil.java:1]: Add header

**Review by zerny — CHANGES_REQUESTED** [review]
Thanks for cleaning this up! Just a header missing and then lgtm

**Comment by sgjesse** [comment]
One final thing. In `third_party/NOTICE` line 67 we have:
```
Various portions of the code are taken from the Android Open Source Project,
and are used in accordance with the following license:
```
Please add the following at line 86 (just before the Apache 2.0 license):
```
Various portions of the code are taken from https://github.com/google/guava,
and are used in accordance with the following license:
```

File: dexlib2/src/main/java/com/android/tools/smali/util/InputStreamUtil.java (hunk 7 above)
Comment (melcz) [reply dexlib2/src/main/java/com/android/tools/smali/util/InputStreamUtil.java:1]: Done!

**Review by sgjesse — APPROVED** [review]

**Review by sgjesse — APPROVED** [review]

**Pull request merged**